[WIP?] Fix #1 and #2

Lucky-Loek opened the pull request:
While fixing #1, I found out that constants in the form of

`self::CONSTANT = VALUE` aren't detected.

I'll add a test case later, so you can see what's happening. Commands are testless, since I have no experience in unit testing commands. Manual testing is working.

There's also some TODOs scattered in the code. Glad to hear your feedback.

Commit:
Add compiler class and necessary dependencies

Commit:
Add compile command for CLI

Commit:
Update phpdoc

Commit:
Add Commands

Commit:
Add executable for commands

Commit:
Update dependencies to keep up with latest commits

Commit:
Update readme

Commit:
Update test case to fail when using constants in same class

Commit:
Remove var_dump

Commit:
Fix bugs where wrong filenames were used

BackEndTea commented:
Thanks for the help, i'll give it a better look later on, for now, could you run `make cs-fix` to automatically fix style errors, and run `make analyze --keep-going` to see what errors php-stan reports, and see if you can fix those.

Commit:
Fix some naming

Commit:
Fix coding style and possible null pointer

Commit:
Fix PHAR generator

Commit:
CS Fix

Lucky-Loek commented:
Fixed what I could. The unit test is deliberately failing so you can see what's happening. Rest is working and I've fixed a couple of bugs that prevented the .phar from running reliably. It does now!

![image](https://user-images.githubusercontent.com/9716643/38877167-6205c61a-425e-11e8-99fa-7fafad76d7be.png)

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+
+    public function compile(string $filename = 'php-optimize.phar'): void
+    {
+        // TODO figure out what php compatibility should be, now it's 7+ due to typehint
```
PHP compatibility is 7.1+

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+ *
+ * @author Loek van der Linde <lind0077@hz.nl>
+ */
+class Compiler
```
I don't have much experience with creation of phars, but what is the reason of creating our own, instead of using something like box?

Lucky-Loek replied:
I have had some bad experiences with Box and how supposedly good forks handle their dependencies. Can always change this, but creating them this way is kinda like second nature nowadays ;)

BackEndTea commented on tests/PhpParser/NodeVisitor/ConvertorVisitorTest.php:
```diff
@@ -51,12 +51,12 @@ public function providesConvertableCode(): \Generator
 
         yield 'It converts the use of a fully classified const' => [
-            ['Foo\Bar\Baz' => new Node\Scalar\String_('A value')],
+            ['Baz' => new Node\Scalar\String_('A value')],
```
Could you revert this change, and create an issue for this bug?

BackEndTea commented on bin/compile:
```diff
@@ -0,0 +1,15 @@
+    echo 'Failed to compile phar: [' . get_class($e) . '] '
+    . $e->getMessage() . ' at ' . $e->getFile() . ':' . $e->getLine() . PHP_EOL;
+    exit(1);
+}
```
Please add a new line at the end of the file

BackEndTea commented on bin/php-optimize:
```diff
@@ -0,0 +1,13 @@
+$application->add(new PhpOptimizer\Command\ConstantToValueCommand());
+$application->add(new PhpOptimizer\Command\RunPhpOptimizerCommand());
+
+$application->run();
```
Please add a new line at the end of the file

BackEndTea commented on src/Command/ConstantToValueCommand.php:
```diff
@@ -0,0 +1,126 @@
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Class ConstantToValue.
```
Please remove the `Class ConstantToValue.` comment.

BackEndTea commented on src/Command/ConstantToValueCommand.php:
```diff
@@ -0,0 +1,126 @@
+
+        $prettyPrinter = new StandardPrettyPrinter();
+
+        // TODO Now we're looping twice, this should be easier to do
```
For now looping twice is fine. Later we can index all the items we want to change and only loop over those again.

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+     * @param \SplFileInfo $file
+     * @param bool         $strip
+     */
+    private function addFile($phar, $file, $strip = true): void
```
```php
 private function addFile(\Phar $phar, \SplFileInfo $file, bool $strip = true): void
```

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return string
+     */
+    private function getRelativeFilePath($file)
```
```php
private function getRelativeFilePath(\SplFileInfo $file): string
```

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return string
+     */
+    private function stripWhitespace($source)
```
```php
private function stripWhitespace(string $source): string
```

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @param \Phar $phar
+     */
+    private function addBinExecutable($phar): void
```
```php
private function addBinExecutable(\Phar $phar): void
```

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return string
+     */
+    private function getStub()
```
```php
private function getStub(): string
```

BackEndTea commented on src/Compiler.php:
```diff
@@ -0,0 +1,204 @@
+            ->exclude('Tests')
+            ->exclude('tests')
+            ->exclude('docs')
+            ->in(__DIR__ . '/../vendor/nikic/php-parser/')
```
Could we jut add  `../vendor`, if we run install the dependencies with not dev?

BackEndTea replied:
On that note, if possible, could you add a make target that runs these steps?

Lucky-Loek replied:
That requires some heavy tinkering, since I've put the necessary libraries for PHAR creation (`PharUtils` and `symfony/finder`) in the --dev corner. I'll take a look at it, but for now I'm skipping it.

BackEndTea reviewed (commented):
It looks pretty good from what i can seen, i'll check it out on a functional level in a bit.

Commit:
Add newline

Commit:
Remove class comment

Commit:
Add typehints and return types

Commit:
Revert test case

Commit:
Add make target for phar creation

Commit:
Update readme

Commit:
Add newline

Commit:
Convert spaces to tabs

Commit:
Remove unnecessary TODO

Commit:
Add high level Compiler test

Commit:
Style updates

Commit:
Update README.md

BackEndTea commented:
Thanks for your help @loekiedepo
